### PR TITLE
feat: Login by password - Issue #10

### PR DESCRIPTION
**Pull request by Cobidex**
## Checklist
- [x] Implemented endpoint to register user
- [x] Validated request body and returned validation errors when validation fails
- [x] Prevented duplicate record by making email unique
- [x] Created user once validation passes
- [x] Generated a jwt token and refresh token for user on successful login
- [x] Returned basic user details as part of response payload
- [x] Returned generated JWT token along with the response payload
- [x] Refactored user entity into the user module
- [x] Created a base class with redundant fields for different entities to inherit from 
- [x] Add test cases for login success and invalid credentials

## Reference Issue
[FEAT]: Authentication - User Log In](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/10)

### CODE REVIEW

**Comment by Heba-WebDev** [comment]
@Cobidex Please link the corresponding issue and follow this pr format: https://github.com/hngprojects/hng_boilerplate_nestjs/pull/341

**Review by Heba-WebDev — APPROVED** [review]
I'm not sure why mongoose was added as a dependency but everything else seems good to me. I had a review call with you so I don't have much to add.

File: src/modules/auth/auth.service.ts
Comment (Heba-WebDev) [inline src/modules/auth/auth.service.ts:1]: The error response, please let it match the example given in the issue.

**Comment by davidolanrewaju** [comment]
Also, please can you move the entities/parent entity folder into the database folder and also wrap user.entity file in a folder labelled entities

File: src/database/data-source.ts [hunk 1]
```diff
@@ -12,7 +13,7 @@ const dataSource = new DataSource({
   password: process.env.DB_PASSWORD,
   host: process.env.DB_HOST,
   database: process.env.DB_DATABASE,
-  entities: [process.env.DB_ENTITIES],
+  entities: [User],
```
Comment (Homoakin619) [inline src/database/data-source.ts:16]: Kindly revert this change to what it previously was

File: src/modules/auth/tests/login.e2e-spec.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    app = moduleFixture.createNestApplication();
+    app.setGlobalPrefix('api/v1');
+    await app.init();
+  });
+
+  afterAll(async () => {
+    await app.close();
+  });
+
+  it('/api/v1/auth/login (POST) - success', async () => {
```
Comment (buka4rill) [inline src/modules/auth/tests/login.e2e-spec.ts:48]: write proper test statements

"it("should login a user", blah blah"

File: src/modules/auth/dto/login.dto.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import { IsEmail, IsNotEmpty, IsString } from 'class-validator';
+
+export class LoginDto {
+  @IsNotEmpty()
+  @IsEmail({}, { message: 'Please enter correct email address' })
```
Comment (buka4rill) [inline src/modules/auth/dto/login.dto.ts:5]: Not sure if the message is required as the validation already does this

File: src/modules/auth/auth.service.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+@Injectable()
+export class AuthService {
+  constructor(
+    @InjectRepository(User)
+    private usersRepository: Repository<User>,
+    private jwtService: JwtService
+  ) {}
+
+  async login(loginDto: LoginDto): Promise<{}> {
```
Comment (buka4rill) [inline src/modules/auth/auth.service.ts:19]: why are you returning a promise of empty object for login? You returns should have a shape, no?

**Review by buka4rill — COMMENTED** [review]
Couple of comments

**Review by buka4rill — COMMENTED** [review]
Rename the PR to 
feat: Login by password - Issue #10 

read the guidelines I gave for this

**Comment by Cobidex** [comment]
Ok @buka4rill working in it

File: src/modules/auth/auth.service.ts (hunk 4 above)
Comment (Cobidex) [reply src/modules/auth/auth.service.ts]: Issue resolved

File: src/modules/auth/dto/login.dto.ts (hunk 3 above)
Comment (Cobidex) [reply src/modules/auth/dto/login.dto.ts]: Removed

File: src/modules/auth/tests/login.e2e-spec.ts (hunk 2 above)
Comment (Cobidex) [reply src/modules/auth/tests/login.e2e-spec.ts]: Resolved

**Comment by Cobidex** [comment]
@buka4rill issue resolved. Please review

**Comment by buka4rill** [comment]
resolve conflicts

File: src/modules/auth/auth.service.ts (hunk 4 above)
Comment (buka4rill) [reply src/modules/auth/auth.service.ts]: Still see it in the code tho
